move gateway certs specified in server TLS settings to SDS

ramaraochavali opened the pull request:
Fixes https://github.com/istio/istio/issues/25128

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | cb49ea7c6c4c196920642ac0a754dd6c5d54629a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/release-notes_istio/503) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
/test release-notes_istio

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | cb49ea7c6c4c196920642ac0a754dd6c5d54629a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/17364) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
unit-tests_istio | cb49ea7c6c4c196920642ac0a754dd6c5d54629a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/17364) | `/test unit-tests_istio`
integ-multicluster-k8s-tests_istio | cb49ea7c6c4c196920642ac0a754dd6c5d54629a | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-multicluster-k8s-tests_istio/7397) | `/test integ-multicluster-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

hzxuzhonghu commented on pilot/pkg/networking/core/v1alpha3/gateway.go:
```diff
@@ -423,44 +423,22 @@ func buildGatewayListenerTLSContext(
-					MatchSubjectAltNames: util.StringToExactMatch(server.Tls.SubjectAltNames),
-				},
-			}
+	case server.Tls.Mode == networking.ServerTLSSettings_MUTUAL || server.Tls.Mode == networking.ServerTLSSettings_SIMPLE:
```
Add a default branch

hzxuzhonghu replied:
It seems this branch is not equal to previous

ramaraochavali replied:
ok. changed the case to default

ramaraochavali commented:
>  but it conflicts with what we decided about moving gateway SDS to Istiod.
Sorry. I was not aware of this decision. Closing this

costinm commented:
BTW - if you have extra time and are interested in SDS - I would appreciate some help with someone taking over https://github.com/istio/istio/pull/25438, moving the gateway SDS in Istiod. I'm a bit too busy right now.

istio-testing commented:
@ramaraochavali: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

istio-testing commented:
@ramaraochavali: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`
integ-telemetry-mc-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-mc-k8s-tests_istio/1341632341239402496) | `/test integ-telemetry-mc-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`
integ-telemetry-mc-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-mc-k8s-tests_istio/1341632341239402496) | `/test integ-telemetry-mc-k8s-tests_istio`
integ-pilot-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-pilot-k8s-tests_istio/1341632341138739200) | `/test integ-pilot-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`
integ-telemetry-mc-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-mc-k8s-tests_istio/1341632341239402496) | `/test integ-telemetry-mc-k8s-tests_istio`
integ-pilot-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-pilot-k8s-tests_istio/1341632341138739200) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-security-k8s-tests_istio/1341632341176487936) | `/test integ-security-k8s-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`
integ-telemetry-mc-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-mc-k8s-tests_istio/1341632341239402496) | `/test integ-telemetry-mc-k8s-tests_istio`
integ-pilot-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-pilot-k8s-tests_istio/1341632341138739200) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-security-k8s-tests_istio/1341632341176487936) | `/test integ-security-k8s-tests_istio`
release-test_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/release-test_istio/1341632341100990464) | `/test release-test_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@ramaraochavali: The following tests **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integ-distroless-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-distroless-k8s-tests_istio/1341632341289734144) | `/test integ-distroless-k8s-tests_istio`
integ-telemetry-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-k8s-tests_istio/1341632341205848064) | `/test integ-telemetry-k8s-tests_istio`
unit-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/unit-tests_istio/1341632341050658816) | `/test unit-tests_istio`
integ-telemetry-mc-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-telemetry-mc-k8s-tests_istio/1341632341239402496) | `/test integ-telemetry-mc-k8s-tests_istio`
integ-pilot-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-pilot-k8s-tests_istio/1341632341138739200) | `/test integ-pilot-k8s-tests_istio`
integ-security-k8s-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-security-k8s-tests_istio/1341632341176487936) | `/test integ-security-k8s-tests_istio`
release-test_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/release-test_istio/1341632341100990464) | `/test release-test_istio`
integ-operator-controller-tests_istio | bc6e99eafdf8bcb98deba5df9118a461dba8c189 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/25786/integ-operator-controller-tests_istio/1341632341340065792) | `/test integ-operator-controller-tests_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

ramaraochavali commented:
@howardjohn As discussed, rebased and changed. PTAL.

@costinm Discussed with @howardjohn offline. Since DR certs (which are file based), use file based SDS, we need to run SDS at agent event at gateway. Only credential names can use SDS at Istiod. So this PR is still relevant. Please take a look.

howardjohn reviewed (commented):
Do we have existing integration tests for this?

ramaraochavali commented:
The existing gateway certs should cover this case and there is an integration test for file mounted certs delivered via SDS.

ramaraochavali commented:
@costinm Please take a look when you get chance

ramaraochavali commented:
@costinm Can you PTAL when you get chance and if you are OK, can  you remove the "changes requested"?
